Enable poweroff feature on NXP MCXA and MCXN series MCUs.

ZhaoxiangJin opened the pull request:
This PR enabled poweroff feature for NXP MCXA and MCXN series MCUs.

ZhaoxiangJin committed:
modules: hal_nxp: Pulling cmc, vbat, wuu driver

Pulling MCUX SDK cmc, vbat, wuu driver to Zephyr build tree.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: nxp: mcx/mcxa: Enable poweroff for mcxa series

This commit introduced the following changes:
1. Selected kconfig option 'HAS_POWEROFF' for NXP
   MCXA series.
2. Added power.c for NXP MCXA series, in this file,
   we currently implemented 'z_sys_poweroff()' function.
3. For the MCXA series, after waking up from the deep
   power down mode, the reset handler will be executed,
   and we need to release the I/O pads and certain peripheral
   devices to normal operating mode in 'soc_reset_hook'.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: mcxa series: Enable FRO16k for LPTMR

Enabled FRO16k for LPTMR when LPTMR clock source
is set to 0x1.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
samples: nxp/mcx: Enable poweroff for MCXA/N series

This commit enabled poweroff case for NXP MCXA and
MCXN series MCUs.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

butok commented:
[samples/boards/nxp/mcx/mcxa/poweroff/src/main.c](https://github.com/zephyrproject-rtos/zephyr/pull/92844/files#diff-8c4581714797811593476b9667166d334018278523e8605594f6c1689c90dd8e) contains the common Zephyr API code.

Can we avoid creating the NXP board-specific samples in the `samples/boards/nxp/mcx/mcxa/poweroff/` and instead place them in a common location, such as `samples/subsys/poweroff` or another common sample directory?

ZhaoxiangJin committed:
boards/nxp: mcxn: Enable LPUART pin internal pullup resistor

This commit enables MCXN236, MCXN947 LPUART pin internal
pullup resistor. For MCXN947 and MCXN236, during LPUAR
initialization, the RX pin is pulled down internally
and STAT [RAF] is set to one. In this state, attempting
to enter low power mode will trigger LPACK reset and
therefore cannot truly enter low power mode. The correct
setting should be to enable LPUART pin internal pullup resistor.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
soc: nxp: mcx/mcxn: Enable poweroff for mcxn series

This commit introduced the following changes:
1. Selected kconfig option 'HAS_POWEROFF' for NXP
   MCXN series.
2. Added power.c for NXP MCXN series, in this file,
   we currently implemented 'z_sys_poweroff()' function.
3. For the MCXN series, after waking up from the deep
   power down mode, the reset handler will be executed,
   and we need to release the I/O pads and certain peripheral
   devices to normal operating mode in 'soc_reset_hook'.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

ZhaoxiangJin committed:
boards: nxp: mcxn series: Enable FRO16k for LPTMR

Enabled FRO16k for LPTMR when LPTMR clock source
is set to 0x1.

Signed-off-by: Zhaoxiang Jin <Zhaoxiang.Jin_1@nxp.com>

dleach02 commented:
@butok it does make sense. The yaml could be conditioned on the HAS_POWEROFF. 

@hakehuang, can you provide guidance?

ZhaoxiangJin commented:
> @butok it does make sense. The yaml could be conditioned on the HAS_POWEROFF.
> 
> @hakehuang, can you provide guidance?

How about create a NXP MCUs general poweroff demo under samples/boards/nxp/poweroff?

hakehuang commented:
> In my opinion, vendor-specific samples break the main Zephyr philosophy, and they should be an exception (probably, approved by TSC).

@butok , you can take a look at the tests/subsys/pm which is much more well organized, and I am working on a power shield to fully support this. As to NPI enablement, maybe it is ok to add a board vendor demo first for now, we need catchup step by step.
